docs/source/tutorial/kernel.rst

```diff
@@ -263,16 +263,128 @@ Accessing texture (surface) memory in :class:`~cupy.RawKernel` is supported via
+
+.. note::
+    To be able to directly use ``std::complex<float>`` and ``std::complex<double>``, you need to include the header ``<cuda/std/complex.h>`` which is provided by `NVIDIA C++ Standard Library` version 1.4.0 and above. Currently **this header is not shipped with CUDA toolkit 11.2**, see this `link <https://nvidia.github.io/libcudacxx/releases.html>`_ to get more information.
+
+.. _custom_user_structs:
+
+Custom user types
+-----------------
+
+Is is possible to use custom types (composite types such as structures and structures of structures) 
+as kernel arguments by defining a custom numpy data type.
+When doing this, it is your responsability to match host and device structure memory layout.
+The CUDA standard guarantees that the size of fundamental types on the host and device always match.
+It may however impose device alignment requirements on composite types.
+This means that for composite types, the struct member offsets may be different from what you might expect.
+
+When a kernel argument is passed by value, the CUDA driver will copy exactly ``sizeof(param_type)`` bytes starting from the beginning of the numpy object data pointer where ``param_type`` is the parameter type in your kernel. 
+You have to match ``param_type`` memory layout (size, alignment and struct padding/packing) 
+by defining a corresponding `numpy dtype <https://numpy.org/doc/stable/reference/arrays.dtypes.html>`_.
+
+For builtin cuda vector types such as ``int2`` and ``double4`` and other packed structures with 
+named members you can directly define such numpy datatype as the following:
+
+.. doctest::
+
+    >>> import numpy as np
+    >>> names = ['x', 'y', 'z']
+    >>> types = [np.float32]*3
+    >>> float3 = np.dtype({'names': names, 'formats': types})
+    >>> arg = np.random.rand(3).astype(np.float32).view(float3)
+    >>> print(arg)
+    [(0.9940819, 0.62873816, 0.8953669)]
+    >>> arg['x'] = 42.0
+    >>> print(arg)
+    [(42., 0.62873816, 0.8953669)]
+
+Here ``arg`` can be used directly as a kernel argument.
+When there is no need to name fields you may prefer this syntax to define packed structures such as 
+vectors or matrices:
+
+.. doctest::
+
+    >>> import numpy as np
+    >>> float5x5 = np.dtype({'names': ['dummy'], 'formats': [(np.float32,(5,5))]}) 
+    >>> arg = np.random.rand(25).astype(np.float32).view(float5x5)
+    >>> print(arg.itemsize)
+    100
+
+Here ``arg`` represents a 100 bytes scalar (i.e. a numpy array of size 1)
+that can be passed by value to any kernel.
+Upper bound for total kernel parameters size is 256 bytes for compute capability `1.x` (via shared memory)
+and 4kB for compute capability `2.x` and higher (via constant memory).
+
+Composite types can be built recursively using numpy dtype `offsets` and `itemsize` capabilities,
+see ``cupy/tests/cupy_tests/core_tests/test_function.py`` for examples of advanced usage where structures offsets are determined by calling a cuda kernel.
```

Comment:
I wonder if it's better to showcase a few examples in [`cupy/examples`](https://github.com/cupy/cupy/tree/master/examples) instead of referring users to the test suite (although I also did this at times), which contains some noise from the test framework. But we can also do this in a separate PR IMHO.